File: src/main/java/linkedlist/MergeKSortedLists.java
package linkedlist;

import structs.ListNode;

public class MergeKSortedLists {
    public static void main(String[] args){
        MergeKSortedLists s = new MergeKSortedLists();
        ListNode head = new ListNode(1, new ListNode(4, new ListNode(5)));
        ListNode head1 = new ListNode(1, new ListNode(3, new ListNode(4)));
        ListNode res = s.mergeKLists(new ListNode[]{head, head1});
        while(res != null){
            System.out.print(res.val + "->");
            res = res.next;
        }
        System.out.print("\n");
    }

    public ListNode mergeKLists(ListNode[] lists) {
        ListNode dummy = new ListNode(-1);
        ListNode copy = dummy;
        for(ListNode listNode : lists){
            dummy.next = listNode;
            while(dummy.next != null){
                dummy = dummy.next;
            }
        }
        return sort(copy.next);
    }

    private ListNode sort(ListNode head){
        if(head == null || head.next == null){
            return head;
        }
        ListNode middle = findMidNode(head);
        ListNode rightHalf = middle.next;
        middle.next = null;
        return merge(sort(head), sort(rightHalf));
    }

    private ListNode findMidNode(ListNode node){
        ListNode slow = node, fast = node.next;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }

    private ListNode merge(ListNode head1, ListNode head2){
        ListNode dummy = new ListNode(-1);
        ListNode copy = dummy;
        while(head1 != null && head2 != null){
            if(head1.val < head2.val){
                dummy.next = head1;
                head1 = head1.next;
            } else {
                dummy.next = head2;
                head2 = head2.next;
            }
            dummy = dummy.next;
        }

        if(head1 != null){
            dummy.next = head1;
        } else {
            dummy.next = head2;
        }
        return copy.next;
    }
}

File: src/MergeKSortedLists.kt
import structs.ListNode

fun main() {
    val l1: ListNode = ListNode(1, ListNode(4, ListNode(5)))
    val l2: ListNode = ListNode(1, ListNode(3, ListNode(4)))
    val l3: ListNode = ListNode(2, ListNode(6))
    var head: ListNode? = mergeKLists(arrayOf(l1, l2, l3))
    while (head != null) {
        print("${head.`val`} ")
        head = head.next
    }
    print("\n")
}

fun mergeKLists(lists: Array<ListNode?>): ListNode? {
    var dummy: ListNode? = ListNode(-1)
    val copy: ListNode? = dummy
    for (node in lists) {
        dummy?.next = node
        while (dummy?.next != null) {
            dummy = dummy.next
        }
    }
    return sort(copy?.next)
}

fun sort(node: ListNode?): ListNode? {
    if (node == null || node.next == null) {
        return node
    }

    val midNode: ListNode? = findMidNode(node)
    val right: ListNode? = midNode?.next
    midNode?.next = null
    return merge(sort(node), sort(right))
}

fun merge(l1: ListNode?, l2: ListNode?): ListNode? {
    var dummy: ListNode? = ListNode(-1)
    val copy: ListNode? = dummy
    var h1: ListNode? = l1
    var h2: ListNode? = l2
    while (h1 != null && h2 != null) {
        if (h1.`val` < h2.`val`) {
            dummy?.next = h1
            h1 = h1.next
        } else {
            dummy?.next = h2
            h2 = h2.next
        }
        dummy = dummy?.next
    }
    if (h1 != null) {
        dummy?.next = h1
    } else {
        dummy?.next = h2
    }
    return copy?.next
}

fun findMidNode(node: ListNode?): ListNode? {
    if (node == null || node.next == null) {
        return node
    }

    var slow: ListNode? = node
    var fast: ListNode? = node.next
    while (fast != null && fast.next != null) {
        slow = slow?.next
        fast = fast.next?.next
    }
    return slow
}